File: solutions/python/sliding-window/sliding-window-maximum.py
# LeetCode 239 — Sliding Window Maximum (Hard)
# Category: Sliding Window · Approach: Monotonic Deque
# Time: O(n) | Space: O(k)
# Source: https://leetcode.com/problems/sliding-window-maximum/

from collections import deque


def max_sliding_window(nums: list[int], k: int) -> list[int]:
    dq = deque()
    res = []
    for i, x in enumerate(nums):
        while dq and dq[0] <= i - k:
            dq.popleft()
        while dq and nums[dq[-1]] <= x:
            dq.pop()
        dq.append(i)
        if i >= k - 1:
            res.append(nums[dq[0]])
    return res

File: solutions/typescript/sliding-window/sliding-window-maximum.ts
// LeetCode 239 — Sliding Window Maximum (Hard)
// Category: Sliding Window · Approach: Monotonic Deque
// Time: O(n) | Space: O(k)
// Source: https://leetcode.com/problems/sliding-window-maximum/

function maxSlidingWindow(nums: number[], k: number): number[] {
  const res: number[] = [];
  const dq: number[] = [];
  for (let i = 0; i < nums.length; i++) {
    while (dq.length && dq[0] <= i - k) dq.shift();
    while (dq.length && nums[dq[dq.length - 1]] <= nums[i]) dq.pop();
    dq.push(i);
    if (i >= k - 1) res.push(nums[dq[0]]);
  }
  return res;
}
